src/function/scalar/list/list_zip.cpp

```diff
@@ -0,0 +1,157 @@
+#include "duckdb/common/types/data_chunk.hpp"
+#include "duckdb/function/scalar/nested_functions.hpp"
+#include "duckdb/planner/expression_binder.hpp"
+#include "duckdb/planner/expression/bound_function_expression.hpp"
+#include "duckdb/planner/expression/bound_parameter_expression.hpp"
+
+namespace duckdb {
+
+static void ListZipFunction(DataChunk &args, ExpressionState &state, Vector &result) {
+	idx_t count = args.size();
+	idx_t args_size = args.ColumnCount();
+	auto *result_data = FlatVector::GetData<list_entry_t>(result);
+	auto &result_struct = ListVector::GetEntry(result);
+	auto &struct_entries = StructVector::GetEntries(result_struct);
+	bool flags_given = false;
+
+	// Check flag
+	if (args.data.back().GetType().id() == LogicalTypeId::BOOLEAN) {
+		flags_given = true;
+		args_size--;
+	}
+
+	vector<UnifiedVectorFormat> input_lists;
+	for (idx_t i = 0; i < args.ColumnCount(); i++) {
+		UnifiedVectorFormat curr;
+		args.data[i].ToUnifiedFormat(count, curr);
+		input_lists.push_back(curr);
+	}
+
+	// Handling output row for each input row
+	idx_t result_size = 0;
+	vector<idx_t> lengths;
+	for (idx_t j = 0; j < count; j++) {
+
+		// Is flag for current row set
+		bool is_flag = false;
+		if (flags_given) {
+			idx_t flag_idx = input_lists.back().sel->get_index(j);
+			auto flag_data = UnifiedVectorFormat::GetData<bool>(input_lists.back());
+			is_flag = flag_data[flag_idx];
+		}
+
+		// Calculation of the outgoing list size
+		idx_t len = is_flag ? INT_MAX : 0;
+		for (idx_t i = 0; i < args_size; i++) {
+			idx_t curr_size;
+			if (args.data[i].GetType() == LogicalType::SQLNULL || ListVector::GetListSize(args.data[i]) == 0) {
+				curr_size = 0;
+			} else {
+				idx_t sel_idx = input_lists[i].sel->get_index(j);
+				auto curr_data = UnifiedVectorFormat::GetData<list_entry_t>(input_lists[i]);
+				curr_size = input_lists[i].validity.RowIsValid(sel_idx) ? curr_data[sel_idx].length : 0;
+			}
+
+			// Dependent on flag using gt or lt
+			if (is_flag) {
+				len = len > curr_size ? curr_size : len;
+			} else {
+				len = len < curr_size ? curr_size : len;
+			}
+		}
+		lengths.push_back(len);
+		result_size += len;
+	}
+
+	ListVector::SetListSize(result, result_size);
+	ListVector::Reserve(result, result_size);
+
+	idx_t offset = 0;
+	for (idx_t j = 0; j < count; j++) {
```

Comment:
As a remark, this is copying elements at a time again. Again, we can do a pass to build selection vectors instead and then slice the lists into the result struct by slicing their respective children. But I realize I already requested many changes, so let's focus on that first.

Reply:
Ok, I have nevertheless made the implementation with selectionvectoren. I hope that fits so.